### PR TITLE
feature: add support for per-service endpoint url configuration

### PR DESCRIPTION
**Pull request by stobrien89**
*Issue #, if available:*
https://github.com/aws/aws-sdk-php/issues/1239

*Description of changes:*
Adds the ability to configure endpoints on a per-service basis via the environment and ini file.  Order of precedence is as follows:

- `endpoint` client configuration
- `AWS_ENDPOINT_URL_<SERVICE_IDENTIFIER>` env var
- `services` endpoint in shared config  example: 
```
[default]
region = us-west-2
services = my-services

[services my-services]
s3 =
  endpoint_url = https://s3-endpoint-override.aws
```
- `AWS_ENDPOINT_URL` env var
- `endpoint_url` in shared config.  example:
```
[default]
endpoint_url = http://localhost:1234
```

For backward-compatibility, we will continue to support the `endpoint` client constructor option, however, `endpoint_url` will take precedence, if resolved from any one of the above-mentioned options.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/ClientResolver.php [hunk 1]
```diff
@@ -287,6 +287,20 @@ class ClientResolver
...
+            'type'      => 'value',
+            'valid'     => ['bool'],
+            'doc'       => 'Set to true to disable configured endpoint urls.',
+            'fn'        => [__CLASS__, '_apply_ignore_configured_endpoint_urls'],
+            'default'   => [__CLASS__, '_default_ignore_configured_endpoint_urls'],
+        ],
+        'endpoint_url' => [
+            'type'      => 'value',
+            'valid'     => ['string'],
+            'doc'       => 'The full URI of the webservice. This is only required when connecting to a custom endpoint (e.g., a local version of S3). This will take precedence over the `endpoint` configuration option.',
```
Comment (SamRemis) [inline src/ClientResolver.php:300]: This doc could be a bit clearer about how this differs from endpoint

File: src/AwsClient.php [hunk 2]
```diff
@@ -554,7 +554,11 @@ private function setClientBuiltIns($args)
         $config = $this->getConfig();
         $service = $args['service'];
 
-        $builtIns['SDK::Endpoint'] = isset($args['endpoint']) ? $args['endpoint'] : null;
+        $builtIns['SDK::Endpoint'] = isset($config['endpoint_url'])
```
Comment (SamRemis) [inline src/AwsClient.php:557]: Should we merge this soon so we can add this to the deprecation with new syntax?  Like `$config['endpoint_url'] ?? $args['endpoint'] ?? null;`

File: src/ClientResolver.php [hunk 3]
```diff
@@ -287,6 +287,20 @@ class ClientResolver
             'default'   => false,
             'fn'        => [__CLASS__, '_apply_suppress_php_deprecation_warning']
         ],
+        'ignore_configured_endpoint_urls' => [
+            'type'      => 'value',
+            'valid'     => ['bool'],
+            'doc'       => 'Set to true to disable configured endpoint urls.',
```
Comment (SamRemis) [inline src/ClientResolver.php:293]: endpoint urls that have been configured where?

Comment (stobrien89) [reply src/ClientResolver.php:293]: Yeah, could use some clarification here.

File: src/ClientResolver.php (hunk 1 above)
Comment (stobrien89) [reply src/ClientResolver.php:300]: It doesn't really differ from `endpoint`.  It's essentially the same thing with different a different name and configuration locations.  I can add that though.

File: src/Configuration/ConfigurationResolver.php [hunk 4]
```diff
@@ -30,7 +30,8 @@ public static function resolve(
         $key,
         $defaultValue,
         $expectedType,
-        $config = []
+        $config = [],
```
Comment (SamRemis) [inline src/Configuration/ConfigurationResolver.php:33]: Aren't config and options _basically_ synonyms?  Why are we differentiating them here?  Seems like we could flatten this into one array

File: src/Configuration/ConfigurationResolver.php [hunk 5]
```diff
@@ -89,8 +96,13 @@ public static function env($key, $expectedType)
...
      * @return null | mixed
      */
-    public static function ini($key, $expectedType, $profile = null, $filename = null)
-    {
+    public static function ini(
+        $key,
+        $expectedType,
+        $profile = null,
+        $filename = null,
+        $options = []
```
Comment (SamRemis) [inline src/Configuration/ConfigurationResolver.php:104]: Let's stick with config if we are going to stick with config

File: src/functions.php [hunk 6]
```diff
@@ -503,6 +503,42 @@ function boolean_value($input)
...
+    foreach ($lines as $line) {
+        $line = trim($line);
+
+        if (empty($line) || strpos($line, ';') === 0 || strpos($line, '#') === 0) {
+            continue; // Ignore empty lines and comments
+        }
+
+        if (strpos($line, '[') === 0) {
+            $current_service = '';
+            $section_start = (trim($line, '[]') === $section_name);
```
Comment (SamRemis) [inline src/functions.php:523]: So this takes an input, looking for a section name.  Would it make sense to move the logic around a bit?  Maybe make an explicit iterator, and then look for the right section and do the parsing from there until we hit the end of the section?

File: src/Configuration/ConfigurationResolver.php (hunk 4 above)
Comment (stobrien89) [reply src/Configuration/ConfigurationResolver.php:33]: `config` is passed in from the client args/config in the client constructor, but I suppose we could flatten additional options into it.

**Review by SamRemis — APPROVED** [review]

File: src/ClientResolver.php [hunk 7]
```diff
@@ -281,6 +281,20 @@ class ClientResolver
...
             'default'   => true,
         ],
+        'ignore_configured_endpoint_urls' => [
+            'type'      => 'value',
+            'valid'     => ['bool'],
+            'doc'       => 'Set to true to disable endpoint urls configured using the `endpoint_url` configuration option.',
+            'fn'        => [__CLASS__, '_apply_ignore_configured_endpoint_urls'],
+            'default'   => [__CLASS__, '_default_ignore_configured_endpoint_urls'],
+        ],
+        'endpoint_url' => [
```
Comment (SamRemis) [inline src/ClientResolver.php:291]: Since this does the same thing as endpoint, why not just have endpoint look in the config variables and environment variables?

File: src/ClientResolver.php [hunk 8]
```diff
@@ -1116,6 +1130,64 @@ public static function _default_signing_region(array &$args)
...
+            return;
+        }
+
+        $args['config']['endpoint_url'] = $value;
+    }
+
+    public static function _default_endpoint_url(array &$args)
+    {
+        if ($args['config']['ignore_configured_endpoint_urls']
+            || !self::isValidService($args['service'])
```
Comment (SamRemis) [inline src/ClientResolver.php:1160]: This line is just for tests?

File: src/functions.php [hunk 9]
```diff
@@ -503,6 +503,51 @@ function boolean_value($input)
     return null;
 }
 
+/**
+ * Checks if an input is a valid epoch time
```
Comment (SamRemis) [inline src/functions.php:507]: This doesn't seem accurate

File: src/Configuration/ConfigurationResolver.php [hunk 10]
```diff
@@ -100,6 +114,20 @@ public static function ini($key, $expectedType, $profile = null, $filename = nul
         // Use INI_SCANNER_NORMAL instead of INI_SCANNER_TYPED for PHP 5.5 compatibility
         //TODO change after deprecation
         $data = @\Aws\parse_ini_file($filename, true, INI_SCANNER_NORMAL);
+
+        if (isset($options['section'])
+            && isset($options['subsection'])
+            && isset($options['key']))
+        {
+            return self::retrieveValueFromIniSubsection(
```
Comment (SamRemis) [inline src/Configuration/ConfigurationResolver.php:122]: As this is expected to grow and add more config options to individual services, does it make more sense to grab the subsection early on in the client resolver instead and reference it here and have that be referenced throughout the resolution of the client?

File: src/functions.php [hunk 11]
```diff
@@ -503,6 +503,51 @@ function boolean_value($input)
...
+ * @param $filename
+ * @return array
+ */
+function parse_ini_section_with_subsections($filename, $section_name) {
+    $config = [];
+    $ini_content = file_get_contents($filename);
+    $lines = explode("\n", $ini_content);
+
+    $section_start = false;
+    $current_subsection = '';
```
Comment (SamRemis) [inline src/functions.php:519]: Since we know what section we are looking for, would it make sense to instead search the file straight for the section name (using a regex maybe?), then parse each line until the end of the section

Comment (SamRemis) [reply src/functions.php:519]: If not a regex, then we know the exact string we are looking for, maybe just checking if the line equals that with a trim(), then once we have found it using another loop to parse each value into the config array until we are done

File: src/ClientResolver.php (hunk 8 above)
Comment (stobrien89) [reply src/ClientResolver.php:1160]: More or less. There are other places in the clientresolver where we don't require a service or set a default value when a service is not valid/present.

File: src/ClientResolver.php (hunk 7 above)
Comment (stobrien89) [reply src/ClientResolver.php:291]: To ensure consistency in naming across all sources.

Comment (SamRemis) [reply src/ClientResolver.php:291]: Alternate option, we could add an alias field and a little logic in the resolver code so that it gets taken care of there

File: src/functions.php [hunk 12]
```diff
@@ -503,6 +503,63 @@ function boolean_value($input)
...
+    $stream = fopen($filename, 'r');
+
+    if (!$stream) {
+        return $config;
+    }
+
+    $current_subsection = '';
+
+    while (!feof($stream)) {
+        $line = trim(fgets($stream));
```
Comment (SamRemis) [inline src/functions.php:524]: should we condense these few lines into another function called read_ini_line?  Just to avoid the duplication of the trim, the fgets, and the empty/comment check

**Review by SamRemis — APPROVED** [review]

File: src/ClientResolver.php [hunk 13]
```diff
@@ -993,6 +1001,11 @@ public static function _apply_user_agent($inputUserAgent, array &$args, HandlerL
 
     public static function _apply_endpoint($value, array &$args, HandlerList $list)
     {
+        if (empty($value)) {
```
Comment (SamRemis) [inline src/ClientResolver.php:1004]: Can we move this down in resolution order instead?

File: src/CloudSearchDomain/CloudSearchDomainClient.php [hunk 14]
```diff
@@ -35,6 +36,7 @@ public static function getArguments()
             // (e.g. http://search-blah.{region}.cloudsearch.amazonaws.com)
             return explode('.', new Uri($args['endpoint']))[1];
         };
+        unset($args['endpoint']['default']);
```
Comment (SamRemis) [inline src/CloudSearchDomain/CloudSearchDomainClient.php:39]: Why is this necessary?  This seems like an endpoint is always required, so it should never go to the default?

**Review by SamRemis — APPROVED** [review]

File: src/CloudSearchDomain/CloudSearchDomainClient.php (hunk 14 above)
Comment (stobrien89) [reply src/CloudSearchDomain/CloudSearchDomainClient.php:39]: The client resolver won't handle the 'required' property if there's a default
